File: node_java/src/RaftNode.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.*;
import java.util.concurrent.*;

public class RaftNode {
    private String nodeId;
    private int port;
    private List<Peer> peers;
    private String state = "FOLLOWER";
    private int currentTerm = 0;
    private String votedFor = null;
    private long lastHeartbeat;
    private long electionTimeout;
    private boolean running = true;
    private List<String> votesReceived = new ArrayList<>();
    private List<String> log = new ArrayList<>();
    private Map<String, AIModel> models = new HashMap<>();

    private static class Peer {
        String host;
        int port;

        Peer(String host, int port) {
            this.host = host;
            this.port = port;
        }
    }

    public RaftNode(String nodeId, int port, List<Peer> peers) {
        this.nodeId = nodeId;
        this.port = port;
        this.peers = peers;
        loadState();
        resetElectionTimeout();
        this.lastHeartbeat = System.currentTimeMillis();
    }

    private synchronized void saveState() {
        try {
            StringBuilder sb = new StringBuilder();
            sb.append("{");
            sb.append("\"currentTerm\":").append(currentTerm).append(",");
            sb.append("\"votedFor\":").append(votedFor == null ? "null" : "\"" + votedFor + "\"").append(",");
            sb.append("\"log\":[");
            for (int i = 0; i < log.size(); i++) {
                sb.append("\"").append(log.get(i).replace("\"", "\\\"")).append("\"");
                if (i < log.size() - 1)
                    sb.append(",");
            }
            sb.append("],\"models\":{");
            int count = 0;
            for (String mid : models.keySet()) {
                AIModel m = models.get(mid);
                sb.append("\"").append(mid).append("\":{")
                        .append("\"type\":\"").append(m.getType()).append("\",")
                        .append("\"weights\":\"").append(m.serialize()).append("\"}")
                        .append(++count < models.size() ? "," : "");
            }
            sb.append("}}");
            Files.write(Paths.get("raft_state_" + nodeId + ".json"), sb.toString().getBytes());
        } catch (Exception e) {
            System.err.println("Error guardando estado: " + e.getMessage());
        }
    }

    private void loadState() {
        try {
            String path = "raft_state_" + nodeId + ".json";
            if (Files.exists(Paths.get(path))) {
                String content = new String(Files.readAllBytes(Paths.get(path)));
                currentTerm = Integer.parseInt(getJsonValue(content, "currentTerm", "0"));
                votedFor = getJsonValue(content, "votedFor");
                if ("null".equals(votedFor))
                    votedFor = null;

                // Carga de log simplificada (buscando el array [])
                int logStart = content.indexOf("\"log\":[") + 7;
                int logEnd = content.lastIndexOf("]");
                if (logStart > 7 && logEnd > logStart) {
                    String logContent = content.substring(logStart, logEnd);
                    if (!logContent.isEmpty()) {
                        String[] items = logContent.split("\",\"");
                        for (String item : items) {
                            log.add(item.replace("\"", "").replace("\\", ""));
                        }
                    }
                }
                System.out.println("[" + nodeId + "] Estado cargado: Term " + currentTerm + ", Votado por " + votedFor
                        + ", Log " + log.size() + " items");

                // Carga de modelos
                int modelsStart = content.indexOf("\"models\":{");
                if (modelsStart != -1) {
                    // Muy simplificado: buscamos los bloques de modelos
                    // En un sistema real usaríamos Jackson/Gson
                    String modelsSection = content.substring(modelsStart + 9, content.length() - 1);
                    if (modelsSection.length() > 2) {
                        String[] modelEntries = modelsSection.split("\\},\"");
                        for (String entry : modelEntries) {
                            try {
                                String cleanEntry = entry.startsWith("{") ? entry.substring(1) : entry;
                                String mId = cleanEntry.split("\":\\{")[0].replace("\"", "");
                                String mType = getJsonValue(entry, "type");
                                String mWeights = getJsonValue(entry, "weights");

                                AIModel model;
                                if ("MLP".equals(mType)) {
                                    model = new AIModel.MLP(mId, 2, 5);
                                } else {
                                    model = new AIModel.Perceptron(mId, 2);
                                }
                                model.deserialize(mWeights);
                                synchronized (this) {
                                    models.put(mId, model);
                                }
                            } catch (Exception ex) {
                            }
                        }
                    }
                }
            }
        } catch (Exception e) {
            System.err.println("Error cargando estado: " + e.getMessage());
        }
    }

    private void resetElectionTimeout() {
        this.electionTimeout = 3000 + new Random().nextInt(2000); // 3-5 seconds
    }

    public void start() {
        // 1. Thread del Servidor Raft
        new Thread(this::startServer).start();

        // 2. Thread del Monitor Web
        int webPort = port + 2001;
        new Thread(() -> startWebMonitor(webPort)).start();

        // 3. Thread del Election Timer
        new Thread(this::runElectionTimer).start();

        // 4. Thread del Lider (Heartbeats)
        Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(this::runLeaderLoop, 0, 2, TimeUnit.SECONDS);

        System.out
                .println("[" + nodeId + "] Nodo iniciado. Port: " + port + ", Web: " + webPort + ", Estado: " + state);
    }

    private synchronized void processMessage(String json, PrintWriter out) {
        // Parsing manual ultra-básico para evitar librerías pesadas
        String type = getJsonValue(json, "type");
        int term = Integer.parseInt(getJsonValue(json, "term", "0"));

        if (term > currentTerm) {
            currentTerm = term;
            state = "FOLLOWER";
            votedFor = null;
            saveState();
        }

        if ("HEARTBEAT".equals(type)) {
            if (term >= currentTerm) {
                if (term > currentTerm) {
                    currentTerm = term;
                    saveState();
                }
                state = "FOLLOWER";
                lastHeartbeat = System.currentTimeMillis();
            }
            if (out != null)
                out.println("{\"type\": \"HEARTBEAT_ACK\", \"term\": " + currentTerm + "}");
        } else if ("VOTE_REQUEST".equals(type)) {
            String candidateId = getJsonValue(json, "candidate_id");
            boolean canVote = (votedFor == null || votedFor.equals(candidateId)) && (term >= currentTerm);

            if (canVote) {
                votedFor = candidateId;
                currentTerm = term;
                lastHeartbeat = System.currentTimeMillis();
                saveState();
                if (out != null)
                    out.println("{\"type\": \"VOTE_RESPONSE\", \"term\": " + currentTerm + ", \"vote_granted\": true}");
            } else {
                if (out != null)
                    out.println(
                            "{\"type\": \"VOTE_RESPONSE\", \"term\": " + currentTerm + ", \"vote_granted\": false}");
            }
        } else if ("TRAIN".equals(type)) {
            // Permitir entrenamiento si somos LIDER O si recibimos una orden del LIDER
            // (delegación)
            if (!"LEADER".equals(state) && term < currentTerm) {
                if (out != null)
                    out.println("{\"type\": \"TRAIN_RESULT\", \"status\": \"error\", \"message\": \"Not the leader\"}");
                return;
            }
            String modelId = getJsonValue(json, "model_id");
            String modelType = getJsonValue(json, "model_type", "PERCEPTRON");

            // Extraer datos personalizados o usar default
            double[][] inputs;
            double[] targets;

            String customInputs = getJsonValue(json, "inputs");
            String customTargets = getJsonValue(json, "targets");

            if (customInputs != null && customTargets != null) {
                String[] rows = customInputs.replace("[[", "").replace("]]", "").split("\\],\\s*\\[");
                inputs = new double[rows.length][];
                for (int i = 0; i < rows.length; i++) {
                    String[] cols = rows[i].split(",");
                    inputs[i] = new double[cols.length];
                    for (int j = 0; j < cols.length; j++)
                        inputs[i][j] = Double.parseDouble(cols[j].trim());
                }
                String[] tVals = customTargets.replace("[", "").replace("]", "").split(",");
                targets = new double[tVals.length];
                for (int i = 0; i < tVals.length; i++)
                    targets[i] = Double.parseDouble(tVals[i].trim());
            } else {
                inputs = new double[][] { { 0, 0 }, { 0, 1 }, { 1, 0 }, { 1, 1 } };
                targets = new double[] { 0, 1, 1, 1 };
            }

            AIModel finalModel;
            int inputSize = inputs[0].length;

            if (peers.isEmpty()) {
                if ("MLP".equals(modelType))
                    finalModel = new AIModel.MLP(modelId, inputSize, 5);
                else
                    finalModel = new AIModel.Perceptron(modelId, inputSize);
                finalModel.train(inputs, targets, 1000, 0.1);
            } else {
                System.out.println("[" + nodeId + "] Distribuyendo entrenamiento entre " + peers.size() + " nodos...");
                int chunkSize = (int) Math.ceil((double) inputs.length / (peers.size() + 1));
                List<AIModel> subModels = new ArrayList<>();

                // 1. Local
                int localEnd = Math.min(chunkSize, inputs.length);
                double[][] localIn = java.util.Arrays.copyOfRange(inputs, 0, localEnd);
                double[] localTarget = java.util.Arrays.copyOfRange(targets, 0, localEnd);
                AIModel localModel;
                if ("MLP".equals(modelType))
                    localModel = new AIModel.MLP(modelId, inputSize, 5);
                else
                    localModel = new AIModel.Perceptron(modelId, inputSize);
                localModel.train(localIn, localTarget, 1000, 0.1);
                subModels.add(localModel);

                // 2. Peers
                for (int i = 0; i < peers.size(); i++) {
                    int start = (i + 1) * chunkSize;
                    if (start >= inputs.length)
                        break;
                    int end = Math.min(start + chunkSize, inputs.length);
                    double[][] peerIn = java.util.Arrays.copyOfRange(inputs, start, end);
                    double[] peerTarget = java.util.Arrays.copyOfRange(targets, start, end);

                    String taskMsg = "{\"type\": \"TRAIN_TASK\", \"model_id\": \"" + modelId +
                            "\", \"model_type\": \"" + modelType +
                            "\", \"inputs\": \"" + java.util.Arrays.deepToString(peerIn).replace(" ", "") +
                            "\", \"targets\": \"" + java.util.Arrays.toString(peerTarget).replace(" ", "") + "\"}";

                    Peer p = peers.get(i);
                    String resp = sendAndReceive(p.host, p.port, taskMsg);
                    if (resp != null) {
                        String w = getJsonValue(resp, "weights");
                        AIModel pm;
                        if ("MLP".equals(modelType))
                            pm = new AIModel.MLP(modelId, inputSize, 5);
                        else
                            pm = new AIModel.Perceptron(modelId, inputSize);
                        pm.deserialize(w);
                        subModels.add(pm);
                    }
                }

                // 3. Average
                if ("MLP".equals(modelType)) {
                    finalModel = AIModel.MLP.average(modelId,
                            subModels.stream().map(m -> (AIModel.MLP) m).toArray(AIModel.MLP[]::new));
                } else {
                    finalModel = AIModel.Perceptron.average(modelId,
                            subModels.stream().map(m -> (AIModel.Perceptron) m).toArray(AIModel.Perceptron[]::new));
                }
            }

            models.put(modelId, finalModel);
            String weights = finalModel.serialize();
            String syncMsg = "{\"type\": \"MODEL_SYNC\", \"model_id\": \"" + modelId +
                    "\", \"model_type\": \"" + modelType + "\", \"weights\": \"" + weights + "\"}";
            log.add("TRAINED_DISTRIBUTED: " + modelId);
            for (Peer p : peers)
                new Thread(() -> sendMessage(p.host, p.port, syncMsg)).start();
            saveState();
            if (out != null)
                out.println("{\"type\": \"TRAIN_RESULT\", \"model_id\": \"" + modelId + "\", \"status\": \"success\"}");
        } else if ("TRAIN_TASK".equals(type)) {
            // Un seguidor recibe una parte del entrenamiento
            String modelId = getJsonValue(json, "model_id");
            String modelType = getJsonValue(json, "model_type", "PERCEPTRON");
            String customInputs = getJsonValue(json, "inputs");
            String customTargets = getJsonValue(json, "targets");

            // Parsear inputs/targets (copiado de TRAIN)
            String[] rows = customInputs.replace("[[", "").replace("]]", "").split("\\],\\s*\\[");
            double[][] inputs = new double[rows.length][];
            for (int i = 0; i < rows.length; i++) {
                String[] cols = rows[i].split(",");
                inputs[i] = new double[cols.length];
                for (int j = 0; j < cols.length; j++)
                    inputs[i][j] = Double.parseDouble(cols[j].trim());
            }
            String[] tVals = customTargets.replace("[", "").replace("]", "").split(",");
            double[] targets = new double[tVals.length];
            for (int i = 0; i < tVals.length; i++)
                targets[i] = Double.parseDouble(tVals[i].trim());

            AIModel model;
            int inputSize = inputs[0].length;
            if ("MLP".equals(modelType)) {
                model = new AIModel.MLP(modelId, inputSize, 5);
            } else {
                model = new AIModel.Perceptron(modelId, inputSize);
            }

            model.train(inputs, targets, 1000, 0.1);

            if (out != null) {
                out.println("{\"type\": \"TRAIN_TASK_RESULT\", \"model_id\": \"" + modelId +
                        "\", \"weights\": \"" + model.serialize() + "\"}");
            }
            return;
        } else if ("PREDICT".equals(type)) {
            String modelId = getJsonValue(json, "model_id");
            System.out.println("[" + nodeId + "] PREDICCION solicitada para: " + modelId);
            AIModel model = models.get(modelId);
            if (model == null) {
                System.out.println("[" + nodeId + "] Error: Modelo " + modelId + " no encontrado.");
                if (out != null)
                    out.println(
                            "{\"type\": \"PREDICT_RESULT\", \"status\": \"error\", \"message\": \"Model not found\"}");
                return;
            }
            // Parsear inputs
            String dataStr = getJsonValue(json, "data");
            System.out.println("[" + nodeId + "] Inputs recibidos: " + dataStr);
            String[] parts = dataStr.replace("[", "").replace("]", "").split(",");
            double[] inputs = new double[parts.length];
            for (int i = 0; i < parts.length; i++)
                inputs[i] = Double.parseDouble(parts[i].trim());

            double prediction = model.predict(inputs);
            System.out.println("[" + nodeId + "] Resultado prediccion: " + prediction);
            if (out != null)
                out.println("{\"type\": \"PREDICT_RESULT\", \"model_id\": \"" + modelId + "\", \"prediction\": "
                        + prediction + "}");
        } else if ("MODEL_SYNC".equals(type)) {
            String modelId = getJsonValue(json, "model_id");
            String modelType = getJsonValue(json, "model_type");
            String weights = getJsonValue(json, "weights");

            System.out.println(
                    "[" + nodeId + "] Recibiendo sincronizacion de modelo: " + modelId + " (" + modelType + ")");

            AIModel m;
            try {
                if ("MLP".equals(modelType)) {
                    // hiddenSize;bOutput;bHidden...
                    int hiddenSize = Integer.parseInt(weights.split(";")[0]);
                    // Necesitamos inputSize. Lo inferimos de la parte final de los pesos.
                    // wHidden está al final.
                    String[] parts = weights.split(";");
                    int totalWHidden = parts[4].split(",").length;
                    int inputSize = totalWHidden / hiddenSize;
                    m = new AIModel.MLP(modelId, inputSize, hiddenSize);
                } else {
                    // bias;w1,w2...
                    int inputSize = weights.split(";")[1].split(",").length;
                    m = new AIModel.Perceptron(modelId, inputSize);
                }
                m.deserialize(weights);
                models.put(modelId, m);
                System.out.println("[" + nodeId + "] Modelo " + modelId + " sincronizado correctamente.");
                log.add("REPLICATED_MODEL: " + modelId);
                saveState();
                if (out != null)
                    out.println("{\"type\": \"MODEL_SYNC_ACK\", \"status\": \"success\"}");
            } catch (Exception e) {
                System.out.println("[" + nodeId + "] Error sincronizando modelo " + modelId + ": " + e.getMessage());
                if (out != null)
                    out.println("{\"type\": \"MODEL_SYNC_ACK\", \"status\": \"error\", \"message\": \"" + e.getMessage()
                            + "\"}");
            }
        }
    }

    private void runElectionTimer() {
        while (running) {
            try {
                Thread.sleep(500);
                synchronized (this) {
                    if (!"LEADER".equals(state)) {
                        long elapsed = System.currentTimeMillis() - lastHeartbeat;
                        if (elapsed > electionTimeout) {
                            System.out
                                    .println("[" + nodeId + "] Timeout! Iniciando eleccion Term " + (currentTerm + 1));
                            startElection();
                        }
                    }
                }
            } catch (InterruptedException e) {
            }
        }
    }

    private synchronized void startElection() {
        state = "CANDIDATE";
        currentTerm++;
        votedFor = nodeId;
        votesReceived.clear();
        votesReceived.add(nodeId);
        lastHeartbeat = System.currentTimeMillis();
        resetElectionTimeout();
        saveState();

        for (Peer peer : peers) {
            new Thread(() -> {
                String msg = "{\"type\": \"VOTE_REQUEST\", \"term\": " + currentTerm + ", \"candidate_id\": \"" + nodeId
                        + "\"}";
                String response = sendAndReceive(peer.host, peer.port, msg);
                if (response != null && response.contains("VOTE_RESPONSE")) {
                    synchronized (this) {
                        if ("CANDIDATE".equals(state)
                                && Integer.parseInt(getJsonValue(response, "term", "0")) == currentTerm) {
                            if ("true".equals(getJsonValue(response, "vote_granted"))) {
                                votesReceived.add(peer.host + ":" + peer.port);
                                if (votesReceived.size() > (peers.size() + 1) / 2) {
                                    System.out.println("[" + nodeId + "] Gane la eleccion! Ahora soy el LIDER.");
                                    state = "LEADER";
                                }
                            }
                        }
                    }
                }
            }).start();
        }
    }

    private void runLeaderLoop() {
        synchronized (this) {
            if (!"LEADER".equals(state))
                return;
            lastHeartbeat = System.currentTimeMillis(); // Líder reinicia su propio contador al enviar
        }
        for (Peer peer : peers) {
            String msg = "{\"type\": \"HEARTBEAT\", \"term\": " + currentTerm + ", \"leader_id\": \"" + nodeId + "\"}";
            sendMessage(peer.host, peer.port, msg);
        }
    }

    private void startServer() {
        try (ServerSocket serverSocket = new ServerSocket(port)) {
            while (running) {
                Socket client = serverSocket.accept();
                new Thread(() -> {
                    try (BufferedReader in = new BufferedReader(new InputStreamReader(client.getInputStream()));
                            PrintWriter out = new PrintWriter(client.getOutputStream(), true)) {
                        String line = in.readLine();
                        if (line != null) {
                            processMessage(line, out);
                        }
                    } catch (Exception e) {
                    }
                }).start();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private String sendAndReceive(String host, int port, String json) {
        try (Socket socket = new Socket(host, port);
                PrintWriter out = new PrintWriter(socket.getOutputStream(), true);
                BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
            socket.setSoTimeout(1000);
            out.println(json);
            return in.readLine();
        } catch (Exception e) {
            return null;
        }
    }

    private void sendMessage(String host, int port, String json) {
        try (Socket socket = new Socket(host, port);
                PrintWriter out = new PrintWriter(socket.getOutputStream(), true)) {
            out.println(json);
        } catch (Exception e) {
        }
    }

    // Helper manual para JSON
    private String getJsonValue(String json, String key) {
        return getJsonValue(json, key, null);
    }

    private String getJsonValue(String json, String key, String defaultValue) {
        String search = "\"" + key + "\":";
        int index = json.indexOf(search);
        if (index == -1)
            return defaultValue;

        int start = index + search.length();
        // Saltar espacios y dos puntos
        while (start < json.length() && (json.charAt(start) == ' ' || json.charAt(start) == ':')) {
            start++;
        }

        if (start >= json.length())
            return defaultValue;

        char firstChar = json.charAt(start);
        int end = start;

        if (firstChar == '"') {
            // Es un string: leer hasta la siguiente comilla no escapada
            start++; // saltar comilla inicial
            end = start;
            while (end < json.length() && json.charAt(end) != '"') {
                if (json.charAt(end) == '\\')
                    end++; // saltar escape
                end++;
            }
        } else if (firstChar == '[' || firstChar == '{') {
            // Es array u objeto: leer hasta balancear corchetes/llaves
            int balance = 0;
            char open = firstChar;
            char close = (open == '[') ? ']' : '}';
            while (end < json.length()) {
                if (json.charAt(end) == open)
                    balance++;
                if (json.charAt(end) == close)
                    balance--;
                end++;
                if (balance == 0)
                    break;
            }
        } else {
            // Es un número o boolean: leer hasta coma o fin de objeto
            while (end < json.length() && json.charAt(end) != ',' && json.charAt(end) != '}'
                    && json.charAt(end) != ']') {
                end++;
            }
        }

        if (start >= end)
            return defaultValue;
        return json.substring(start, end).trim();
    }

    private void startWebMonitor(int webPort) {
        try (ServerSocket serverSocket = new ServerSocket(webPort)) {
            while (true) {
                try (Socket client = serverSocket.accept()) {
                    BufferedReader in = new BufferedReader(new InputStreamReader(client.getInputStream()));
                    in.readLine();

                    final String htmlOutput;
                    synchronized (this) {
                        StringBuilder sb = new StringBuilder();
                        sb.append("<html><head><meta http-equiv='refresh' content='2'>")
                                .append("<style>body{font-family:monospace;background:#222;color:#0f0;padding:20px;}")
                                .append(".card{border:1px solid #444;padding:20px;border-radius:8px;max-width:600px;}")
                                .append(".LEADER{color:#f00;} .FOLLOWER{color:#0f0;}.CANDIDATE{color:#ff0;}</style></head>")
                                .append("<body><div class='card'>")
                                .append("<h1>Monitor nodo: ").append(nodeId).append("</h1>")
                                .append("<p>Estado: <span class='").append(state).append("'>").append(state)
                                .append("</span></p>")
                                .append("<p>Termino: ").append(currentTerm).append("</p>")
                                .append("<p>Voto por: ").append(votedFor).append("</p>")
                                .append("<p>")
                                .append("LEADER".equals(state) ? "Latido enviado hace: " : "Ultimo latido hace: ")
                                .append(String.format("%.2f", (System.currentTimeMillis() - lastHeartbeat) / 1000.0))
                                .append("s</p>")
                                .append("<hr><h3>Modelos Entrenados:</h3><ul>");

                        for (String mid : models.keySet()) {
                            sb.append("<li><b>").append(mid).append("</b> (").append(models.get(mid).getType())
                                    .append(")</li>");
                        }

                        sb.append("</ul><hr><p>Log size: ").append(log.size()).append("</p></div></body></html>");
                        htmlOutput = sb.toString();
                    }

                    String response = "HTTP/1.1 200 OK\r\nContent-Type: text/html\r\nContent-Length: "
                            + htmlOutput.length()
                            + "\r\n\r\n" + htmlOutput;
                    client.getOutputStream().write(response.getBytes());
                } catch (Exception e) {
                    // Ignorar errores de conexión individual
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) {
        String id = args.length > 0 ? args[0] : "JavaNode_1";
        int port = args.length > 1 ? Integer.parseInt(args[1]) : 6000;
        List<Peer> peers = new ArrayList<>();
        if (args.length > 2) {
            for (int i = 2; i < args.length; i++) {
                peers.add(new Peer("127.0.0.1", Integer.parseInt(args[i])));
            }
        } else {
            peers.add(new Peer("127.0.0.1", 5000)); // Default PythonNode
        }
        new RaftNode(id, port, peers).start();
    }
}

File: node_python/src/raft_node.py
import socket
import json
import threading
import time
import random

# Estados de Raft
FOLLOWER = "FOLLOWER"
CANDIDATE = "CANDIDATE"
LEADER = "LEADER"

class RaftNode:
    def __init__(self, node_id, port, peers):
        self.node_id = node_id
        self.port = port
        self.peers = peers # Lista de tuplas (ip, port) de los otros nodos
        
        # Estado persistente (Debería guardarse en disco, por ahora en memoria)
        self.current_term = 0
        self.voted_for = None
        self.log = []

        # Estado volátil
        self.state = FOLLOWER
        self.last_heartbeat = time.time()
        self.election_timeout = random.uniform(3.0, 5.0) 
        self.running = True
        self.lock = threading.Lock()
        self.votes_received = set()

    def start(self):
        # 1. Hilo del Servidor Raft
        server_thread = threading.Thread(target=self.start_server, daemon=True)
        server_thread.start()

        # 2. Hilo del Timeout (Election Timer)
        timer_thread = threading.Thread(target=self.run_election_timer, daemon=True)
        timer_thread.start()

        # 3. Hilo del Servidor Web (Monitor)
        self.web_port = self.port + 3000
        web_thread = threading.Thread(target=self.start_web_monitor, daemon=True)
        web_thread.start()

        # 4. Hilo del Lider (Solo activo si es LIDER)
        leader_thread = threading.Thread(target=self.run_leader_loop, daemon=True)
        leader_thread.start()

        print(f"[{self.node_id}] Nodo iniciado. Raft-Port: {self.port}, Web-Port: {self.web_port}, Estado: {self.state}")
        
        try:
            while self.running:
                time.sleep(1)
        except KeyboardInterrupt:
            print("Cerrando nodo...")

    def start_web_monitor(self):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            s.bind(('0.0.0.0', self.web_port))
            s.listen()
            while self.running:
                try:
                    client, addr = s.accept()
                    request = client.recv(1024).decode('utf-8')
                    
                    with self.lock:
                        html_content = f"""
                        <html>
                        <head>
                            <meta http-equiv="refresh" content="2">
                            <style>
                                body {{ font-family: monospace; background: #222; color: #0f0; padding: 20px; }}
                                .card {{ border: 1px solid #444; padding: 20px; border-radius: 8px; max-width: 600px; }}
                                h1 {{ color: #fff; }}
                                .state {{ font-size: 2em; font-weight: bold; }}
                                .leader {{ color: #f00; }}
                                .follower {{ color: #0f0; }}
                                .candidate {{ color: #ff0; }}
                            </style>
                        </head>
                        <body>
                            <div class="card">
                                <h1>Monitor nodo: {self.node_id}</h1>
                                <p>Estado actual:</p>
                                <div class="state {self.state.lower()}">{self.state}</div>
                                <hr>
                                <p>Termino actual: {self.current_term}</p>
                                <p>Voto por: {self.voted_for}</p>
                                <p>{ "Latido enviado hace" if self.state == LEADER else "Ultimo latido recibido hace" }: {time.time() - self.last_heartbeat:.2f}s</p>
                                <p>Peers: {self.peers}</p>
                                <h3>Log de entradas:</h3>
                                <pre>{json.dumps(self.log, indent=2)}</pre>
                            </div>
                        </body>
                        </html>
                        """
                    
                    response = (
                        "HTTP/1.1 200 OK\r\n"
                        "Content-Type: text/html\r\n"
                        f"Content-Length: {len(html_content)}\r\n"
                        "Connection: close\r\n\r\n"
                        + html_content
                    )
                    client.sendall(response.encode('utf-8'))
                    client.close()
                except Exception as e:
                    pass

    def start_server(self):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            s.bind(('0.0.0.0', self.port))
            s.listen()
            while self.running:
                client, addr = s.accept()
                threading.Thread(target=self.handle_client, args=(client,)).start()

    def handle_client(self, client_sock):
        with client_sock:
            try:
                data = client_sock.recv(16384).decode('utf-8')
                if not data: return
                
                for line in data.split('\n'):
                    if not line.strip(): continue
                    try:
                        msg = json.loads(line)
                        response = self.process_message(msg)
                        if response:
                            client_sock.sendall((json.dumps(response) + "\n").encode('utf-8'))
                    except json.JSONDecodeError:
                        pass
            except Exception as e:
                pass

    def process_message(self, msg):
        msg_type = msg.get("type")
        term = msg.get("term", 0)

        with self.lock:
            # Si recibimos un término mayor, volvemos a follower
            if term > self.current_term:
                self.current_term = term
                self.state = FOLLOWER
                self.voted_for = None
            
            if msg_type == "HEARTBEAT":
                if term >= self.current_term:
                    self.current_term = term
                    self.state = FOLLOWER
                    self.last_heartbeat = time.time()
                return {"type": "HEARTBEAT_ACK", "term": self.current_term}

            elif msg_type == "VOTE_REQUEST":
                candidate_id = msg.get("candidate_id")
                can_vote = (self.voted_for is None or self.voted_for == candidate_id) and (term >= self.current_term)
                
                if can_vote:
                    self.voted_for = candidate_id
                    self.current_term = term
                    self.last_heartbeat = time.time() # Reset timer upon voting
                    return {"type": "VOTE_RESPONSE", "term": self.current_term, "vote_granted": True}
                else:
                    return {"type": "VOTE_RESPONSE", "term": self.current_term, "vote_granted": False}

            elif msg_type == "VOTE_RESPONSE":
                if self.state == CANDIDATE and term == self.current_term:
                    if msg.get("vote_granted"):
                        # Usamos el set de IDs para no contar dos veces si hay reintentos
                        # (Aunque en sockets TCP es raro, es buena practica)
                        # Nota: msg no trae id por defecto en la respuesta, lo asumimos del flujo o lo agregamos
                        pass 
                return None

            elif msg_type == "TRAIN":
                if self.state != LEADER:
                    return {"type": "TRAIN_RESULT", "status": "error", "message": "Not the leader"}
                
                # Mock de entrenamiento (Fase 3 se encargará de esto)
                model_id = msg.get("model_id")
                data = msg.get("data", [])
                loss = sum(x * x for x in data)
                self.log.append(f"Model {model_id} loss: {loss}")
                return {"type": "TRAIN_RESULT", "model_id": model_id, "status": "success", "loss": loss}

        return None

    def run_election_timer(self):
        while self.running:
            time.sleep(0.5)
            with self.lock:
                if self.state != LEADER:
                    elapsed = time.time() - self.last_heartbeat
                    if elapsed > self.election_timeout:
                        print(f"[{self.node_id}] Timeout! Iniciando eleccion para Term {self.current_term + 1}")
                        self.start_election()

    def start_election(self):
        self.state = CANDIDATE
        self.current_term += 1
        self.voted_for = self.node_id
        self.votes_received = {self.node_id}
        self.last_heartbeat = time.time()
        self.election_timeout = random.uniform(3.0, 5.0)

        # Enviar VOTE_REQUEST a todos los peers
        for peer_ip, peer_port in self.peers:
            threading.Thread(target=self.request_vote_from_peer, args=(peer_ip, peer_port)).start()

    def request_vote_from_peer(self, ip, port):
        msg = {
            "type": "VOTE_REQUEST",
            "term": self.current_term,
            "candidate_id": self.node_id
        }
        response = self.send_json(ip, port, msg)
        if response and response.get("type") == "VOTE_RESPONSE":
            with self.lock:
                if self.state == CANDIDATE and response.get("term") == self.current_term:
                    if response.get("vote_granted"):
                        # Como no enviamos ID en la respuesta, simplemente contamos votos únicos por conexión
                        # Para ser más exacto, el mensaje de respuesta debería incluir el peer_id
                        self.votes_received.add(f"{ip}:{port}")
                        
                        if len(self.votes_received) > (len(self.peers) + 1) / 2:
                            print(f"[{self.node_id}] Gane la eleccion! Ahora soy el LIDER.")
                            self.state = LEADER

    def run_leader_loop(self):
        while self.running:
            if self.state == LEADER:
                with self.lock:
                    self.last_heartbeat = time.time() # El lider reinicia su propio contador al enviar
                # Enviar Heartbeats
                for peer_ip, peer_port in self.peers:
                    threading.Thread(target=self.send_heartbeat, args=(peer_ip, peer_port)).start()
            time.sleep(2)

    def send_heartbeat(self, ip, port):
        msg = {
            "type": "HEARTBEAT",
            "term": self.current_term,
            "leader_id": self.node_id
        }
        self.send_json(ip, port, msg)

    def send_json(self, ip, port, msg):
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                s.settimeout(1.0)
                s.connect((ip, port))
                s.sendall((json.dumps(msg) + "\n").encode('utf-8'))
                # Esperar respuesta opcional
                data = s.recv(1024).decode('utf-8')
                if data:
                    return json.loads(data.split('\n')[0])
        except:
            return None
        return None

if __name__ == "__main__":
    # Configuración para pruebas locales
    import sys
    node_id = sys.argv[1] if len(sys.argv) > 1 else "PythonNode_1"
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 5000
    
    # Peers: Si corremos varios en local, pasamos los otros puertos
    peers = []
    if len(sys.argv) > 3:
        for p in sys.argv[3:]:
            peers.append(('127.0.0.1', int(p)))
    else:
        # Por defecto asume que hay un JavaNode en 6000
        peers = [('127.0.0.1', 6000)]
        
    node = RaftNode(node_id, port, peers)
    node.start()

